FIX-#2322: add aligning partition' blocks

anmyachev opened the pull request:
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/CONTRIBUTING.html
if you have questions about contributing.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

The problem is that the situation when modin frames have the same number of rows, but different blocks (`partition.list_of_blocks`) were not processed. It is wrong to get block sizes based on the index only, since in general, several partitions are suitable for one index.

Example of incorrect partitioning:

```python
(Pdb)  [obj.length() for obj in self._partitions.T[0]]
[56, 56, 56, 52]
(Pdb)  [obj.length() for obj in self._partitions.T[1]]
[55, 55, 55, 55]
```

The blocks after `_copartition` call must be the same, as there is code that relies on this (for example: `_row_lengths` and  `_column_widths`). To do this, I **directly pass the size of the partition' blocks**.

Found and **fixed** another bugs:
* we cannot determine the number of values that the ray remote function `deploy_ray_func` will return to us, based only on `num_splits` parameter, since there is also `_lengths` parameter that affects this.

Found and **reported** another bugs:
* we shouldn't directly change `self._partitions`, since after that the modin frame object becomes inconsistent - when `_row_lengths` or  `_column_widths` functions are called, the caching technique is applied, and when changing the property, so we will use cached values (which may be out of date). See https://github.com/modin-project/modin/issues/2374
  ```python
  self._partitions = self._frame_mgr_cls.map_axis_partitions(
    axis, self._partitions, lambda df: df
  )
  ```

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/CONTRIBUTING.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #2322 <!-- issue must be created for each patch -->
- [x] tests added and passing

dchigarev commented on modin/engines/base/frame/partition_manager.py:
```diff
@@ -245,12 +248,23 @@ def broadcast_axis_partitions(
         # may want to line to partitioning up with another BlockPartitions object. Since
         # we don't need to maintain the partitioning, this gives us the opportunity to
         # load-balance the data as well.
+        kw = {
+            "num_splits": num_splits,
+            "other_axis_partition": right_partitions,
+        }
+        if lengths:
+            kw.update(
+                {
+                    "_lengths": lengths,
+                    "manual_partition": True,
+                }
+            )
```
```suggestion
        kw = {
            "num_splits": num_splits,
            "other_axis_partition": right_partitions,
            "manual_partition": bool(lengths),
            "_lengths": lengths,
        }
```

anmyachev replied:
In some cases, the control falls into `deploy_func_between_two_axis_partitions` that does not work correctly if `_lengths` or `manual_partition`  in kwargs.

prutskov commented:
Looks like that similar problem is observed for binary operations. This should be fixed also, I think.

Code to reproduce:
```python
import modin.pandas as pd

data11 = [0, 1]
data12 = [2, 3]

data21 = [0]
data22 = [1, 2, 3]

df1 = pd.DataFrame(data11)
df1 = df1.append(pd.DataFrame(data12))

df2 = pd.DataFrame(data21)
df2 = df2.append(pd.DataFrame(data22))

print(df1 - df2)
```

<details><summary>Outupt</summary>

```
UserWarning: Distributing <class 'list'> object. This may take some time.
UserWarning: Due to https://github.com/pandas-dev/pandas/issues/35092, Pandas ignores sort=False; Modin correctly does not sort.
Traceback (most recent call last):
  File "tester.py", line 15, in <module>
    print(df1 - df2)
  File "/localdisk/aprutsko/modin/modin/pandas/base.py", line 2705, in __str__
    return repr(self)
  File "/localdisk/aprutsko/modin/modin/pandas/dataframe.py", line 183, in __repr__
    result = repr(self._build_repr_df(num_rows, num_cols))
  File "/localdisk/aprutsko/modin/modin/pandas/base.py", line 168, in _build_repr_df
    return self.iloc[indexer]._query_compiler.to_pandas()
  File "/localdisk/aprutsko/modin/modin/pandas/indexing.py", line 725, in __getitem__
    result = super(_iLocIndexer, self).__getitem__(row_lookup, col_lookup, ndim)
  File "/localdisk/aprutsko/modin/modin/pandas/indexing.py", line 297, in __getitem__
    qc_view = self.qc.view(row_lookup, col_lookup)
  File "/localdisk/aprutsko/modin/modin/backends/pandas/query_compiler.py", line 2826, in view
    self._modin_frame.mask(row_numeric_idx=index, col_numeric_idx=columns)
  File "/localdisk/aprutsko/modin/modin/engines/base/frame/data.py", line 515, in mask
    return self.copy()
  File "/localdisk/aprutsko/modin/modin/engines/base/frame/data.py", line 686, in copy
    return self.__constructor__(
  File "/localdisk/aprutsko/modin/modin/engines/base/frame/data.py", line 66, in __init__
    ErrorMessage.catch_bugs_and_request_email(
  File "/localdisk/aprutsko/modin/modin/error_message.py", line 60, in catch_bugs_and_request_email
    raise Exception(
Exception: Internal Error. Please email bug_reports@modin.org with the traceback and command that caused this error.
Row lengths: 5 != 7
```
</details>

prutskov commented on modin/pandas/test/test_io.py:
```diff
@@ -1893,6 +1893,16 @@ def test_to_gbq():
         modin_df.to_gbq("modin.table")
 
 
+def test_internal_error_2322():
```
I think, that name of test should be more clearly to reflect what it is about without visiting github.

anmyachev replied:
Changed.

anmyachev commented:
> Looks like that similar problem is observed for binary operations. This should be fixed also, I think.
> 
> Code to reproduce:
> 
> ```python
> import modin.pandas as pd
> 
> data11 = [0, 1]
> data12 = [2, 3]
> 
> data21 = [0]
> data22 = [1, 2, 3]
> 
> df1 = pd.DataFrame(data11)
> df1 = df1.append(pd.DataFrame(data12))
> 
> df2 = pd.DataFrame(data21)
> df2 = df2.append(pd.DataFrame(data22))
> 
> print(df1 - df2)
> ```
> 
> Outupt

Added appropriate test and fixed.

prutskov commented on modin/test/test_internals.py:
```diff
@@ -0,0 +1,27 @@
+    repr(accm)
+
+
+def test_aligning_blocks_with_duplicated_index():
+    # Same problem as in `test_internal_error_2322` but with duplicated values in index.
+    data11 = [0, 1]
+    data12 = [2, 3]
+
+    data21 = [0]
+    data22 = [1, 2, 3]
+
+    df1 = pd.DataFrame(data11).append(pd.DataFrame(data12))
+    df2 = pd.DataFrame(data21).append(pd.DataFrame(data22))
+
+    repr(df1 - df2)
```
Another observation:
```python
import modin.pandas as pd
import pandas

data11 = [0, 1]
data12 = [2, 3]

data21 = [0]
data22 = [1, 2, 3]

df1 = pd.DataFrame(data11).append(pd.DataFrame(data12))
df2 = pd.DataFrame(data21).append(pd.DataFrame(data22))
df1.columns = df2.columns = ['a']

pdf1 = pandas.DataFrame(data11).append(pandas.DataFrame(data12))
pdf2 = pandas.DataFrame(data21).append(pandas.DataFrame(data22))
pdf1.columns = pdf2.columns = ['a']

print(df1.a - df2) # give result
print(pdf1.a - pdf2) # failed with exception
```

Should we fail with exception in this case for aligning with pandas?

anmyachev replied:
Since this works: `print(pdf1 - pdf2)`, I would suggest that the inconsistency in the behavior of the pandas itself.

Can you create an appropriate issue in pandas to clarify the expected behavior?

prutskov replied:
Yes, corresponded [issue](https://github.com/pandas-dev/pandas/issues/37663) in pandas was created.

anmyachev replied:
Thanks @prutskov .

prutskov commented on .github/workflows/ci.yml:
```diff
@@ -189,6 +189,8 @@ jobs:
         run: python -m pytest modin/config/test
       - shell: bash -l {0}
         run: python -m pytest modin/test/test_envvar_catcher.py
+      - shell: bash -l {0}
+        run: python -m pytest modin/test/test_internals.py
```
Do we need to add this `test_internals.py` in TeamCity testing scope?

YarShev replied:
I think adding this file to testing is redundant. Both tests can be added into existing files.

anmyachev replied:
What file can you suggest? I would prefer these tests to be near.

anmyachev replied:
> Do we need to add this `test_internals.py` in TeamCity testing scope?

Tests from `test-internals` github action job doesn't include in TeamCity testing scope for now, so `test_internals.py` can also exist only in github actions CI.

prutskov replied:
Do you plan to include this file in `TeamCity` testing scope?

anmyachev replied:
Based on the comment above - no, any objections?

prutskov replied:
I haven't any objections for current state of this file.

YarShev replied:
> What file can you suggest? I would prefer these tests to be near.

I would add them into `modin/pandas/test/dataframe/test_iter.py::test___repr`.

anmyachev replied:
But we actually doesn't test repr, it's just the trigger for check consistency of internal axes.

dchigarev replied:
I'm voting up to make separate tests to cover internal logic, but for this approach a have some objections:

1. These tests are engine-specific, the problem exists in described sense only for engines that inherits logic of partitioning from [`BasePandasFrame`](https://github.com/anmyachev/modin/blob/504fd910390255a72a4bd6fde092e43d42b5d8f4/modin/engines/base/frame/data.py#L27). For example, OmniSci backend have different engine and do not inherit that logic, so these tests don't make sense for OmniSci engine. My suggestion is to explicitly mark that we're testing only limited scope of engines here, maybe we can rename the path to that file like: `modin/test/engines/pandas/test_internals.py`

2. We should not use high-level abstractions like `DataFrame` or `Series` and use query compilers or `ModinFrame` instead.  I'm okay with creating frames with high-level API instruments, but I think that since we can extract from created frames `df._query_compiler` and `df._query_compiler._modin_frame` we should use them to recreate bad test case. Also, I think we should use some more advanced logic of checking that everything is okay instead of just `repr`

anmyachev replied:
Moved `test_internals.py` into `modin/test/backend/pandas/`.

As for the second statement, this applies only indirectly to this pull request. Create an appropriate issue for this.

YarShev replied:
I like @dchigarev's suggestion. Regarding 1, [OmnisciOnRayFrame](https://github.com/modin-project/modin/blob/master/modin/experimental/engines/omnisci_on_ray/frame/data.py#L52) is inherited from `BasePandasFrame` as well. So, I would not divide location into `engines` folder for now and just keep location `modin/test/test_internals.py` for these tests. As expanding tests we will be able to divide them into appropriate locations. Regarding 2, I am okay with this. Also, we will need to create an appropriate issue for moving current tests related to internal errors into new location.

anmyachev replied:
I suggest leaving the folder for tests as it is, because one way or another we will have to change the folder for these tests in the future, so let's think about it when we decide exactly how it should look.

So we will quickly add this fix to the master, which users has been waiting for a long time.

YarShev replied:
Okay.

YarShev replied:
One more comment. Why didn't you include tests into [here](https://github.com/modin-project/modin/blob/master/.github/workflows/push.yml)?

prutskov commented on modin/engines/base/frame/data.py:
```diff
@@ -1702,31 +1708,48 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
         right_old_idxes = index_other_obj
 
         is_avoid_reindex = len(joined_index) != len(joined_index.unique()) and axis == 0
```
`is_avoid_reindex ` is used only in one place in current state. Can we use boolean expression directly in [1713](https://github.com/modin-project/modin/pull/2367/files#diff-64aee86078c1a929c18582ab58f1bc06e251bf21fbf10f523ce433f56ab7b8b0R1713)?

anmyachev replied:
Yes, we can. Fixed.

YarShev commented on modin/engines/base/frame/data.py:
```diff
@@ -1689,44 +1693,61 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
         right_old_idxes = index_other_obj
 
-        is_avoid_reindex = len(joined_index) != len(joined_index.unique()) and axis == 0
+        def make_map_func():
```
Why is `lambda df: df` appropriate rather than `lambda df: df.reindex(joined_index, axis=axis)` in some cases?

anmyachev replied:
This behavior was before my changes, I can only say that if we reindex on values that are not unique, we will get an error.

YarShev commented on modin/engines/base/frame/data.py:
```diff
@@ -1689,44 +1693,61 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
         ):
+            reindexed_self = self._partitions
+        else:
+            # aligning index without aligning partition' blocks
```
This phrase is unclear as well, as for me, as partition' blocks are reshuffled as well. On the whole, I would either delete new comments at all or write more descriptive comments.

prutskov reviewed (dismissed):
Thanks @anmyachev, LGTM!

YarShev commented on modin/engines/base/frame/data.py:
```diff
@@ -1689,44 +1693,61 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
             ):
                 reindexed_other = other[i]._partitions
             else:
+                # aligning index with aligning partition' blocks
```
See comment above.

dchigarev commented on modin/engines/base/frame/data.py:
```diff
@@ -1672,12 +1674,14 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
             ):
                 is_aligning_applied = True
+                # we shouldn't modificate modin frame' partitions directly
                 self._partitions = self._frame_mgr_cls.map_axis_partitions(
                     axis, self._partitions, lambda df: df
                 )
```
I know that these changes do not belong to this PR, but just pointing out that we're aligning `self` partitions in a loop, so we're aligning `self` partition `len(other) times`, when it's enough to make this only once, that logic definitely should be changed in the future, because we're getting here significant performance bottleneck

anmyachev replied:
Open issue https://github.com/modin-project/modin/issues/2374 for this.

anmyachev commented on modin/engines/base/frame/data.py:
```diff
@@ -1689,44 +1693,61 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
-            reindexed_self = self._partitions
-        reindexed_other_list = []
 
+        def get_column_widths(partitions):
```
These are the properties of the frame object, and `partitions` a list of numpy arrays.

dchigarev replied:
It's because we're dealing here with partitions that do not have that fields, not with `ModinFrame`. However probably we can wrap them into a frame, not sure that this is the best idea, so maybe some other variants?

anmyachev commented on modin/engines/ray/pandas_on_ray/frame/axis_partition.py:
```diff
@@ -43,7 +44,7 @@ def deploy_axis_func(
             )
             + tuple(partitions),
-            num_returns=num_splits * 3,
+            num_returns=num_splits * 3 if lengths is None else len(lengths) * 3,
```
This parameter affects the number of splits that are created in the `split_result_of_axis_func_pandas` and subsequently affects the number of `deploy_ray_func` return values.

anmyachev commented on modin/engines/base/frame/data.py:
```diff
@@ -1689,44 +1693,61 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
 
-        is_avoid_reindex = len(joined_index) != len(joined_index.unique()) and axis == 0
+        def make_map_func():
+            if len(joined_index) != len(joined_index.unique()) and axis == 0:
```
yes

anmyachev commented:
@dchigarev  @YarShev ready for review.

YarShev reviewed (changes requested):
@anmyachev , the only my concern is about location of `test_internals.py`. The rest of the changes  are LGTM! Regarding #2374, we can consider it in another PR.

anmyachev commented:
> @anmyachev , the only my concern is about location of `test_internals.py`. The rest of the changes are LGTM! Regarding #2374, we can consider it in another PR.

@dchigarev ?

dchigarev reviewed (approved):
LGTM
